cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/proto/ValueFilter.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.report.proto;
+
+import co.cask.cdap.report.util.ReportField;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Represents a filter that checks whether a given value of a field is one of the allowed values and is not one of
+ * the forbidden values.
+ *
+ * @param <T> type of the values
+ */
+public class ValueFilter<T> extends Filter<T> {
+  @Nullable
+  private final List<T> whitelist;
+  @Nullable
+  private final List<T> blacklist;
+
+  public ValueFilter(String fieldName, @Nullable List<T> whitelist, @Nullable List<T> blacklist) {
+    super(fieldName);
+    this.whitelist = whitelist;
+    this.blacklist = blacklist;
+  }
+
+  /**
+   * @return the allowed values of this field, or an empty list if there's no such limit
+   */
+  public List<T> getWhitelist() {
+    return whitelist == null ? Collections.emptyList() : whitelist;
+  }
+
+  /**
+   * @return the disallowed values of this field, or an empty list if there's no such limit
+   */
+  public List<T> getBlacklist() {
+    return blacklist == null ? Collections.emptyList() : blacklist;
+  }
+
+  @Override
+  public List<String> getErrors() {
```

Comment:
instead of returning a list, you can add to the errors list in the super class and rename this method as `addErrors`

Reply:
since this class will be deserialized from json, we cannot guarantee how the list will be deserialized, it might be immutable, so I think it's safer to return a new list instead